src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ObjectTracker.kt

```diff
@@ -23,38 +50,448 @@ interface ObjectTracker {
+
+/**
+ * A typealias for representing an object identifier.
+ *
+ * The unique object id is a 64-bit integer number, constructed from
+ * the object's serial number and its identity hash code.
+ *
+ * @see ObjectTracker
+ */
+typealias ObjectID = Long
+
+/**
+ * Represents an entry for the tracked object.
+ *
+ * @property objectNumber A unique serial number for the object.
+ * @property objectHashCode The identity hash code of the object.
+ * @property objectDisplayNumber The number used in string representation of the object.
+ * @property objectReference A weak reference to the associated object.
+ */
+open class ObjectEntry(
+    val objectNumber: Int,
+    val objectHashCode: Int,
+    val objectDisplayNumber: Int,
+    val objectReference: WeakReference<Any>,
+)
+
+/**
+ * A unique identifier of an object.
+ *
+ * The identifier is a 64-bit integer number, formed by combining its serial number and hash code.
+ * The serial number [ObjectEntry.objectNumber] is stored in the higher 32-bits of the id, while
+ * the identity hash code [ObjectEntry.objectHashCode] is stored in the lower 32-bits.
+ */
+val ObjectEntry.objectId: ObjectID get() =
+    (objectNumber.toLong() shl 32) + objectHashCode.toLong()
+
+/**
+ * Extracts and returns the object number from the given object id.
+ */
+fun ObjectID.getObjectNumber(): Int =
+    (this ushr 32).toInt()
+
+/**
+ * Extracts and returns the identity hash code of the object from the given object id.
+ *
+ * @return The integer hash code derived from the ObjectID.
  */
-internal object StaticObject: Any()
\ No newline at end of file
+fun ObjectID.getObjectHashCode(): Int =
+    this.toInt()
+
+/**
+ * Retrieves the unique serial object number for the given object.
+ *
+ * @param obj the object for which the object number is to be retrieved.
+ * @return the unique object number if the object is registered in the tracker,
+ *   or -1 if no entry is associated with the given object.
+ */
+fun ObjectTracker.getObjectNumber(obj: Any): Int =
+    get(obj)?.objectNumber ?: -1
+
+/**
+ * Retrieves the display number of a given object.
+ *
+ * @param obj the object for which the display number is to be retrieved.
+ * @return the display number of the object if it exists; otherwise, -1 if the object is not registered.
+ */
+fun ObjectTracker.getObjectDisplayNumber(obj: Any): Int =
+    get(obj)?.objectDisplayNumber ?: -1
+
+/**
+ * Enumerates all objects currently tracked by the object tracker,
+ * assigning each object its display number as determined by the object tracker.
+ *
+ * Enumeration is required for the Plugin as we want to see on the diagram if some object was replaced by a new one.
+ * Uses the same a numeration map as the trace reporter via [getObjectDisplayNumber] method, so objects have the
+ * same numbers, as they have in the trace.
+ *
+ * @return a map associating each reachable object with its display number.
+ */
+internal fun ObjectTracker.enumerateAllObjects(): Map<Any, Int> {
```

Comment:
Minor suggestionk: something like the following would seem more idiomatic to me
```kotlin
internal fun ObjectTracker.enumerateAllObjects(): Map<Any, Int> = enumerateObjectEntries()
    .filter { it.objectReference.get() != null }
    .associate { it.objectReference.get()!! to it.objectDisplayNumber }

```

Reply:
Yes, it is better stylistically, but intermediate `filter` would result in additional intermediate list creation (which can be quite large). I propose to leave the loop code for now.